File: Bubilator88/Rendering/FramePublisher.swift
import Foundation

/// One completed 640×400 RGBA frame plus the display metadata the renderer
/// needs to interpret it.
///
/// The metadata travels *with* the frame on purpose. The Metal view used to
/// ask `machine.bus.is400LineMode` at draw time, which after the emulation
/// thread split would be a read of live machine state from the wrong thread —
/// and could disagree with the frame actually being uploaded if the mode
/// changed in between.
nonisolated final class FrameSlot: @unchecked Sendable {
  var pixels: [UInt8]
  var is400LineMode: Bool = false

  init(pixelCount: Int) {
    pixels = [UInt8](repeating: 0, count: pixelCount)
  }
}

/// Hand-off of completed frames from the emulation thread to the renderer.
///
/// Three slots rotate between three roles: one being filled by the producer,
/// one holding the newest completed frame, one held by the consumer while it
/// uploads. No slot is ever touched by two threads at once, so the lock only
/// guards the small bookkeeping — never the pixel copy.
///
/// The producer never blocks: if the renderer has not picked up the previous
/// frame, that frame is simply dropped. Emulation must not be paced by the
/// display.
nonisolated final class FramePublisher: @unchecked Sendable {
  private let lock = NSLock()
  private let pixelCount: Int

  /// Slots nobody is using.
  private var free: [FrameSlot]
  /// Newest completed frame, waiting to be picked up.
  private var ready: FrameSlot?
  /// The slot the consumer currently holds (returned to `free` when it
  /// acquires the next one).
  private var consuming: FrameSlot?

  private var published: Int = 0

  init(pixelCount: Int) {
    self.pixelCount = pixelCount
    free = (0..<3).map { _ in FrameSlot(pixelCount: pixelCount) }
  }

  /// Total frames published since launch. Used for the FPS readout, which
  /// counts emulated frames rather than draws.
  var publishedCount: Int {
    lock.lock()
    defer { lock.unlock() }
    return published
  }

  // MARK: - Producer (emulation thread)

  /// Publish a completed frame.
  ///
  /// The slot is claimed under the lock, filled outside it, then handed over
  /// under the lock again — so a 1 MB copy never blocks the renderer.
  ///
  /// `counted` distinguishes a frame the machine actually produced from one the
  /// main thread rendered while the loop was parked (a debugger step, a redraw
  /// after a mount). Only the former belongs in the FPS readout.
  func publish(pixels: [UInt8], is400LineMode: Bool, counted: Bool = true) {
    lock.lock()
    let slot: FrameSlot
    if let reusable = free.popLast() {
      slot = reusable
    } else if let stale = ready {
      // Renderer is behind and holds one slot: overwrite the frame it has
      // not picked up rather than stalling emulation.
      ready = nil
      slot = stale
    } else {
      lock.unlock()
      return
    }
    lock.unlock()

    slot.pixels.withUnsafeMutableBufferPointer { dst in
      pixels.withUnsafeBufferPointer { src in
        guard let d = dst.baseAddress, let s = src.baseAddress else { return }
        d.update(from: s, count: min(dst.count, src.count))
      }
    }
    slot.is400LineMode = is400LineMode

    lock.lock()
    if let dropped = ready { free.append(dropped) }
    ready = slot
    if counted { published &+= 1 }
    lock.unlock()
  }

  // MARK: - Consumer (renderer)

  /// Take the newest completed frame, or `nil` if none arrived since the last
  /// call. The returned slot stays valid until the next `acquireLatest()`.
  func acquireLatest() -> FrameSlot? {
    lock.lock()
    defer { lock.unlock() }
    guard let next = ready else { return nil }
    ready = nil
    if let previous = consuming { free.append(previous) }
    consuming = next
    return next
  }
}

File: windows/Bubilator88.Windows/FramePublisher.cs
using System;

namespace Bubilator88.Windows;

/// <summary>
/// One completed 640×400 RGBA frame plus the display metadata the presenter
/// needs to interpret it. <see cref="Is400Line"/> travels with the frame
/// because the presenter must not read live machine state, which belongs to
/// the emulation thread and may already describe a later frame. Mirrors the
/// macOS <c>FrameSlot</c>.
/// </summary>
internal sealed class FrameSlot
{
    public readonly byte[] Pixels;
    public bool Is400Line;

    public FrameSlot(int byteCount) => Pixels = new byte[byteCount];
}

/// <summary>
/// Hand-off of completed frames from the emulation thread to the UI thread
/// (mirrors the macOS <c>FramePublisher</c>).
///
/// <para>Three slots rotate between three roles: one being filled by the
/// producer, one holding the newest completed frame, one held by the consumer.
/// No slot is ever touched by two threads at once, so the lock only guards the
/// bookkeeping — never the pixel copy.</para>
///
/// <para>The producer never blocks: if the UI has not picked up the previous
/// frame, that frame is simply dropped. Emulation must not be paced by the
/// display.</para>
/// </summary>
internal sealed class FramePublisher
{
    private readonly object _lock = new();
    private readonly FrameSlot[] _free = new FrameSlot[3];
    private int _freeCount;
    private FrameSlot? _ready;       // newest completed frame, not yet picked up
    private FrameSlot? _consuming;   // held by the consumer until its next acquire
    private long _published;

    public FramePublisher(int byteCount)
    {
        for (int i = 0; i < _free.Length; i++) _free[i] = new FrameSlot(byteCount);
        _freeCount = _free.Length;
    }

    /// <summary>Frames published with <c>counted: true</c> since creation (FPS readout).</summary>
    public long PublishedCount
    {
        get { lock (_lock) return _published; }
    }

    /// <summary>
    /// Publish a completed frame (producer side). <paramref name="counted"/> is
    /// false for a frame rendered outside the loop (e.g. right after a state
    /// load) so it does not inflate the FPS readout.
    /// </summary>
    public void Publish(ReadOnlySpan<byte> pixels, bool is400Line, bool counted = true)
    {
        FrameSlot slot;
        lock (_lock)
        {
            if (_freeCount > 0)
            {
                slot = _free[--_freeCount];
            }
            else if (_ready is not null)
            {
                // The consumer is behind: overwrite the frame it has not
                // picked up rather than stalling emulation.
                slot = _ready;
                _ready = null;
            }
            else
            {
                return;
            }
        }

        pixels[..Math.Min(pixels.Length, slot.Pixels.Length)].CopyTo(slot.Pixels);
        slot.Is400Line = is400Line;

        lock (_lock)
        {
            if (_ready is not null) _free[_freeCount++] = _ready;
            _ready = slot;
            if (counted) _published++;
        }
    }

    /// <summary>
    /// Take the newest completed frame, or null if none arrived since the last
    /// call (consumer side). The slot stays valid until the next acquire.
    /// </summary>
    public FrameSlot? AcquireLatest()
    {
        lock (_lock)
        {
            if (_ready is not { } next) return null;
            _ready = null;
            if (_consuming is not null) _free[_freeCount++] = _consuming;
            _consuming = next;
            return next;
        }
    }
}
